File: frontend/mobile/GdeOni.Mobile.Shared/Utils/CoordinateParser.cs
using System.Globalization;

namespace GdeOni.Mobile.Shared.Utils;

/// <summary>
/// Парсер строковых значений широты / долготы / точности (метры) в double.
/// Принимает и точку, и запятую как десятичный разделитель (на ru-RU
/// клавиатуре по умолчанию запятая). Парсит всегда через InvariantCulture,
/// чтобы поведение не зависело от системной локали.
/// </summary>
public static class CoordinateParser
{
    /// <summary>
    /// Парсит произвольный double. Trim'ит ввод и заменяет ',' на '.'.
    /// Возвращает false, если ввод не парсится.
    /// </summary>
    public static bool TryParseDouble(string? input, out double value)
    {
        var normalized = (input ?? string.Empty).Trim().Replace(',', '.');
        return double.TryParse(
            normalized,
            NumberStyles.Float,
            CultureInfo.InvariantCulture,
            out value);
    }

    /// <summary>
    /// Парсит широту с проверкой диапазона [-90, 90].
    /// </summary>
    public static bool TryParseLatitude(string? input, out double value)
    {
        if (!TryParseDouble(input, out value)) return false;
        return value is >= -90 and <= 90;
    }

    /// <summary>
    /// Парсит долготу с проверкой диапазона [-180, 180].
    /// </summary>
    public static bool TryParseLongitude(string? input, out double value)
    {
        if (!TryParseDouble(input, out value)) return false;
        return value is >= -180 and <= 180;
    }

    /// <summary>
    /// Парсит точность в метрах — должно быть неотрицательным числом.
    /// </summary>
    public static bool TryParseAccuracy(string? input, out double value)
    {
        if (!TryParseDouble(input, out value)) return false;
        return value >= 0;
    }
}

File: frontend/web/src/utils/coordinateParser.ts
/**
 * F5. Парсер координат для ручного ввода. Зеркало
 * GdeOni.Mobile.Shared/Utils/CoordinateParser.cs.
 *
 * Принимает запятую и точку как десятичный разделитель (юзер
 * копирует из карт Google, Yandex, WhatsApp, etc.), trim'ит пробелы,
 * проверяет диапазоны.
 *
 * Будет переиспользован в F8 (создание у могилы) и F15 (правка
 * координат). Пока живёт здесь, в F8 либо вытащу в /utils/, либо
 * переиспользую из этого же файла.
 */

export function tryParseDouble(input: string | null | undefined): number | null {
  if (!input) return null;
  const cleaned = input.trim().replace(',', '.');
  if (cleaned === '') return null;
  const n = Number(cleaned);
  return Number.isFinite(n) ? n : null;
}

export function tryParseLatitude(input: string | null | undefined): number | null {
  const n = tryParseDouble(input);
  if (n === null) return null;
  return n >= -90 && n <= 90 ? n : null;
}

export function tryParseLongitude(input: string | null | undefined): number | null {
  const n = tryParseDouble(input);
  if (n === null) return null;
  return n >= -180 && n <= 180 ? n : null;
}

export function tryParseAccuracy(input: string | null | undefined): number | null {
  const n = tryParseDouble(input);
  if (n === null) return null;
  return n >= 0 ? n : null;
}
